### ISSUE TITLE
How to use and benchmark Halide autoscheduler?

### ISSUE THREAD

**Issue by Rehanchy** [halide/Halide#8432]
Hi Halide developers, 

I was trying to use Halide autoscheduler to generate scheduler for matmul by following the old [tutorials](https://halide-lang.org/tutorials/tutorial_lesson_21_auto_scheduler_generate.html). (btw it's really old, it seems argument last_level_cache_size and balance are no longer in use nowadays)

I found that the schedule produced by autoscheduler is not having great performance, so I wish to see if you guys can help to check whether I'm using the autoscheduler correctly.

My generator (matmul_generator.cpp) looks like this:

`class MatMulGenerator : public Halide::Generator<MatMulGenerator> {
public:
    Input<Buffer<double>> A{"A", 2};  // Input matrix A (m x l)
    Input<Buffer<double>> B{"B", 2};  // Input matrix B (l x n)
    Output<Buffer<double>> C{"C", 2}; // Output matrix C (m x n)

    void generate() {
        Var x("x"), y("y"), k("k");
        Func result("result");
        RDom r(0, A.dim(1).extent());

        result(x, y) = Halide::Expr(0.0);
        result(x, y) += A(x, r.x) * B(r.x, y);
        C(x, y) = result(x, y);
    }
    void schedule() {
        if (using_autoscheduler()) {
            A.set_estimates({{0, 4096}, {0, 4096}});
            B.set_estimates({{0, 4096}, {0, 4096}});
            C.set_estimates({{0, 4096}, {0, 4096}});
        } else {
            C.compute_root();
        }
    }
};

HALIDE_REGISTER_GENERATOR(MatMulGenerator, matmul_generator)
`

Then I'm using these commands to generate the schedule, following the tutorial.
`g++ matmul_generator.cpp /path/to/GenGen.cpp -g -std=c++17 -fno-rtti -I/path/to/halide/include -L/path/to/halide/lib -lHalide -lpthread -ldl -o matmul_generator`

`./matmul_generator -o . -g matmul_generator -f matmul_autoschedule_true -e static_library,h,schedule -p /path/to/halide/lib/libautoschedule_adams2019.so target=host autoscheduler=Adams2019 autoscheduler.parallelism=8`

In another cpp file, I will use this line of code to call the scheduled matrix multiplication.
`matmul_autoschedule_true(A.raw_buffer(), B.raw_buffer(), C.raw_buffer());`

I also have questions about how to benchmark halide autoscheduler's performance on a given kernel, I know that in test/performance/matrix_multiplication.cpp, `out.realize(output);` is called twice, because there will be code generation phase overhead in the first call, and we need to measure halide's performance with the second call.

**To summarize, my questions are**

1. Is my way of using autoscheduler correct?
2. I have a minor concern that when benchmarking halide using the second realize call, the cache is not cold, which may lead to performance overestimation.
3. When using autoscheduler, and call the kernel like this `matmul_autoschedule_true(A.raw_buffer(), B.raw_buffer(), C.raw_buffer());`, does this function contain the code generation phase that could lead to performance underestimation?

Thanks a lot!

**Comment by abadams** [comment]
> [!NOTE]
> **Automated triage:** this issue was flagged by Claude as likely safe to close, for the reason below. The reasoning was checked against the current source, and a second pass tried and failed to refute it, but no human verified this specific text. **If the reason is wrong, please reopen the issue and say so** — some fraction of these are expected to be wrong.

**Reason (CLOSE_STALE):** This is a usage/benchmarking question, not a defect report: no bug is claimed against a specific behavior and there is nothing to fix in the codebase. The one factual claim about the tutorial is wrong — `last_level_cache_size` and `balance` are still live Mullapudi2016 parameters (the reporter was using Adams2019, which simply does not take them).

**Evidence:** `src/autoschedulers/mullapudi2016/AutoSchedule.cpp:57-62` still parses `last_level_cache_size` and `balance`; `tutorial/lesson_21_auto_scheduler_generate.cpp:14,114-115` documents them for Mullapudi2016 only. AOT-compiled entry points contain no codegen, so calling `matmul_autoschedule_true(...)` measures only the pipeline.

Closing as a usage question — please move follow-ups to GitHub Discussions. Short answers: (1) your invocation is correct, but `last_level_cache_size`/`balance` are Mullapudi2016-only parameters and are not used by Adams2019; (2) an AOT-compiled function such as `matmul_autoschedule_true(...)` contains no compilation work at all, so timing the first call is fine — the double `realize()` in the JIT tests is only to skip JIT compilation; (3) use `tools/halide_benchmark.h`, and note that Adams2019 is not expected to match a hand-tuned/BLAS matmul schedule.
